src/lib/kcps-budget-manager.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Krzysztof Frydryk <frydryk.krzysztof@intel.com>
+
+#include <rtos/spinlock.h>
+#include <sof/sof.h>
+#include <stdint.h>
+#include <sof/lib/kcps-budget-manager.h>
+#include <rtos/clk.h>
+#include <sof/sof.h>
+#include <sof/math/numbers.h>
+#ifdef __ZEPHYR__
+#include <zephyr/sys/util.h>
+#endif /* __ZEPHYR__ */
+
+static struct kcps_budget_data kcps_data;
+
+static int request_freq_change(int core, int freq)
+{
+	int current_freq;
+	int selected_freq_id = 0;
+	struct clock_info *clk;
+
+	clk = clocks_get() + core;
+
+	for (int i = 0; i < clk->freqs_num; i++) {
+		if (freq < clk->freqs[i].freq) {
+			selected_freq_id = i;
+			break;
+		}
+	}
+
+	/* don't change clock frequency if already using proper clock */
+	current_freq = clock_get_freq(core);
+	if (clk->freqs[selected_freq_id].freq != current_freq)
+		clock_set_freq(core, freq);
+
+	return 0;
+}
+
+static int max_core_consumption(void)
+{
+	int result = 0;
+
+	for (int i = 0; i < CONFIG_CORE_COUNT; i++)
+		result = MAX(result, kcps_data.kcps_consumtion[i]);
+
+	return result;
+}
+
+int core_kcps_adjust(int core, int kcps_delta)
```

Comment:
What is the expected call model? Is it allowed to call this from any core, or always from core0? I'm guessing as there is locking, the assumption is that any core can call this, right? This would be good to document in the PR.